python/ray/_private/metrics_agent.py

```diff
@@ -187,12 +275,61 @@ def _record_metrics(self, metrics):
                         )
                     else:
                         raise ValueError("Summary is not supported")
-
                     view_data.tag_value_aggregation_data_map[tag_vals] = data
 
-        # Finally, export all the values
+        # Finally, export all the values.
+        # When the view data is exported, they are hard-copied.
         self.view_manager.measure_to_view_map.export(view_data_changed)
 
+    def clean_all_dead_worker_metrics(self):
+        """Clean dead worker's metrics.
+
+        Worker metrics are cleaned up if there was no metrics
+        report for more than `worker_timeout_s`.
+
+        This method has to be periodically called by a caller.
+        """
+        with self._lock:
+            if not self.view_manager:
+                return
+
+            worker_ids_to_clean = []
+            for worker_id_hex, state in self.worker_id_to_state.items():
+                if time.time() - state.last_reported_time > self.worker_timeout_s:
+                    worker_ids_to_clean.append(worker_id_hex)
+
+            for worker_id in worker_ids_to_clean:
+                self._clean_worker_metrics(worker_id)
+
+    def _clean_worker_metrics(self, worker_id_hex: str):
+        assert self._lock.locked()
+
+        if worker_id_hex not in self.worker_id_to_state:
+            return
+
+        state = self.worker_id_to_state[worker_id_hex]
+        state.view_to_owned_tag_vals
+        view_names_changed = set()
+
+        for view_name, tag_vals_to_clean in state.view_to_owned_tag_vals.items():
+            view_data = self._get_mutable_view_data(view_name)
+            for tag_vals in tag_vals_to_clean:
+                if tag_vals in view_data.tag_value_aggregation_data_map:
+                    # If we remove tag_vals from here,
+                    # exporter will stop exporting metrics.
+                    del view_data.tag_value_aggregation_data_map[tag_vals]
+            view_names_changed.add(view_name)
+
+        # We need to re-export the view data so that prometheus
```

Comment:
will this get eventually picked up at the next export if we don't do it eagerly here?

Reply:
no. There's no next export without this because the metrics won't be reported anymore (export only happens when the metrics are received from a component)